Data/Vinyl/Index.hs

```diff
@@ -0,0 +1,111 @@
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE PolyKinds             #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+
+module Data.Vinyl.Index where
+
+import Data.Vinyl.Core
+import Data.Vinyl.Functor
+
+-- | A Peano index provides a constructive proof that the type-level list @rs@
+-- contains the type @r@.
+data RIndex (rs :: [k]) (r :: k) where
+  Z :: RIndex (r ': rs) r
```

Comment:
Would you mind renaming `Z` and `S` to avoid conflicts with `Nat` from `Data.Vinyl.TypeLevel`?